### ISSUE TITLE
Diff upload does not allow forward references to placeholders

### ISSUE THREAD

**Issue by mstn** [openstreetmap/openstreetmap-website#1465]
Hi, I am writing regarding this issue https://github.com/openstreetmap/iD/issues/3208 for iD editor. The cause of this bug is that the server does not allow to reference an entity that is defined later in the same osmChange request. The problem cannot be resolved on the client side when the changeset contains circular dependencies. Hence, I was wondering if the API should handle this case instead. 

ps: You can find an example to reproduce the problem on the original issue.

**Comment by tomhughes** [comment]
There's no circular dependency in that example as far as I can see, iD has just ordered the members wrongly. If it had swapped the last two there would have been no problem.

I'm not sure why you have opened this, as nobody on the iD seems to have suggested there is an API issue or asked you to report it here?

**Comment by zerebubuth** [comment]
I think a patch that would allow relations to appear in any order would be welcome, although in this instance it could have been fixed on the client since it's an ordering issue rather than circular references.

Not handling circular relation references is, to some degree, a feature rather than a bug. Circular references aren't explicitly disallowed, but it would be strongly recommended not to create them or rely upon them.

Firstly, they create a problem when trying to delete either one, as it relies on the other and so has to be "disentangled" before deletion. This is not handled automatically in the upload process, and is something the client has to handle. Most probably don't.

Secondly, recursive or circular relations might not be handled well in other software that consumes OSM data. Or possibly it might be ignored altogether. Either way, it seems unlikely that whatever was intended by the circular relation would be accurately conveyed.

In summary: Circular relation dependencies, while possible, should be avoided.

**Comment by bhousel** [comment]
Hey @tomhughes & @zerebubuth, I did suggest for @mstn to open this issue here first just to get an opinion of whether it was something worth fixing at the API level.

The example given isn't strictly a "circular" relation, but it is an example where the user created the Route Master first, and the the Route children afterwards.  iD assigns temporary keys in sequential order, so it's not unusual for some relations to contain member keys that appear later in the file.

- Master Route (`r-1`), members: (`r-2`, `r-3`)
- Child Route (`r-2`)
- Child Route (`r-3`)

iD could probably take a pass through all the relations and move `r-1` to the end of the file as a workaround, but... I'd really rather see this fixed in the API.

**Comment by tomhughes** [comment]
@mstn It would probably help to have a slightly more meaningful title for the issue rather than copying the original error from iD which isn't even really what you are now complaining about...

**Comment by tomhughes** [comment]
@bhousel It really doesn't do a topological sort of the changes - it just assumes that editing can only create changes in the right order? Pretty sure JOSM gave that up as a bad plan a long time ago...

**Comment by zerebubuth** [comment]
> I'd really rather see this fixed in the API.

What would you do during diff upload to fix this?

Reordering elements to fix creation dependencies seems simple enough: create nodes, create ways, loop over relations creating any which have members that were already created. Doing this outside of a single create block seems like it could disrupt other dependency patterns between deletions and modifications. And solving the whole operation dependency graph seems like it would be complex, if not Hard.

**Comment by bhousel** [comment]
> What would you do during diff upload to fix this?

I assumed it would take one pass over the data to assign primary keys and then a second pass to actually insert the stuff into the database.  But I do understand it's easier to have it work the way it does in a single pass.

If you don't want to change how key assignment happens, thats ok with me - it just means I'll need to change iD to reorder the relations, and also add some checks to prevent users from creating circular relation graphs (we all agree they should be avoided).

**Comment by Zverik** [comment]
In my opinion, reordering relations in an editor is way simpler than on the API side. For example, think of the case the upload is split into mutiple diffs.

**Comment by pnorman** [comment]
I'd rather keep it the way it is. My reason is the replication stream, which can be generally assumed to be broken down into components which are individually valid. This would break that assumption.

**Comment by zerebubuth** [comment]
> My reason is the replication stream, which can be generally assumed to be broken down into components which are individually valid. This would break that assumption.

I don't understand why reordering on the server or the client would necessarily break that assumption. As long as whatever reordering is done before PKs are assigned, then it seems like the order of items in the replication stream would be the same as the order in which the items were committed to the database. Does that sound correct?

**Comment by pnorman** [comment]
Take the example in https://github.com/openstreetmap/iD/issues/3208#issuecomment-281942743

If you took that diff you'd end up with an osmChange for the changeset of

```xml
<osmChange version="0.6" generator="iD">
   <create>
      <node id="1" lon="11.625506992810122" lat="46.866699181636555" version="0" changeset="3">
         <tag k="highway" v="bus_stop" />
      </node>
      <node id="2" lon="11.62686047585252" lat="46.86730122861715" version="0" changeset="3">
         <tag k="highway" v="bus_stop" />
      </node>
      <relation id="1" version="0" changeset="3">
         <member type="node" role="" ref="1" />
         <tag k="type" v="route" />
         <tag k="name" v="AtoB" />
      </relation>
      <relation id="2" version="0" changeset="3">
         <member type="relation" role="" ref="1" />
         <member type="relation" role="" ref="3" />
         <tag k="type" v="route_master" />
         <tag k="name" v="master" />
      </relation>
      <relation id="3" version="0" changeset="3">
         <member type="node" role="" ref="2" />
         <tag k="type" v="route" />
         <tag k="name" v="BtoA" />
      </relation>
   </create>
</osmChange>
```

If you process this all at once, all the references are there, but if you split it up between the relation 2 and 3 creates you get two diffs, the first of which references relations that don't exist.

**Comment by zerebubuth** [comment]
Is it not possible to assign PKs to get the following?

```xml
<osmChange version="0.6" generator="iD">
   <create>
      <node id="1" lon="11.625506992810122" lat="46.866699181636555" version="0" changeset="3">
         <tag k="highway" v="bus_stop" />
      </node>
      <node id="2" lon="11.62686047585252" lat="46.86730122861715" version="0" changeset="3">
         <tag k="highway" v="bus_stop" />
      </node>
      <relation id="1" version="0" changeset="3">
         <member type="node" role="" ref="1" />
         <tag k="type" v="route" />
         <tag k="name" v="AtoB" />
      </relation>
      <relation id="2" version="0" changeset="3">
         <member type="node" role="" ref="2" />
         <tag k="type" v="route" />
         <tag k="name" v="BtoA" />
      </relation>
      <relation id="3" version="0" changeset="3">
         <member type="relation" role="" ref="1" />
         <member type="relation" role="" ref="2" />
         <tag k="type" v="route_master" />
         <tag k="name" v="master" />
      </relation>
   </create>
</osmChange>
```

This can be done within any `<create>` block via the algorithm I gave earlier. More extensive reordering might be possible, although the explosion of different interactions is giving me a headache.

I think it would be good to have a patch for `<create>`-reordering, and it would help to have this in one place (i.e: the API) rather than in every editor, as long as it's not computationally onerous.

A significant downside, in my opinion, is that it makes it much harder to understand or explain what the upload call is doing. At the moment, it seems fairly easy to understand that it processes the elements in the order they appear in the upload. Reordering some kinds of changes and not others adds complexity and requires explanation, making the API more baffling. And some would say it's already plenty baffling enough.

**Comment by pnorman** [comment]
> This can be done within any <create> block via the algorithm I gave earlier. More extensive reordering might be possible, although the explosion of different interactions is giving me a headache.
> 
> I think it would be good to have a patch for <create>-reordering, and it would help to have this in one place (i.e: the API) rather than in every editor, as long as it's not computationally onerous.
> 
> A significant downside, in my opinion, is that it makes it much harder to understand or explain what the upload call is doing. At the moment, it seems fairly easy to understand that it processes the elements in the order they appear in the upload. Reordering some kinds of changes and not others adds complexity and requires explanation, making the API more baffling. And some would say it's already plenty baffling enough.

I would favour keeping this on the client, because the client already has to reorder and combine operations to go from what the user did to an osmChange, and as you said, this adds complexity to the API. Probably significant complexity when we look at all the cases possible.

In this case the original osmChange probably came from

1. create node -1
1. add tags to node -1
1. create node -2
1. add tags to node -2
1. create relation -1
1. add node -1 to relation -1
1. add tags to relation -1
1. create relation -2
1. add relation -1 to relation -2
1. add tags to relation -2
1. create relation -3
1. add node -1 to relation -3
1. add tags to relation -3
1. add relation -3 to relation -2

When iD turned it into XML it brought step 14 before step 11.

**Comment by mmd-osm** [comment]
I'm closing the issue now as per previous comment 4.5 years ago.
